### ISSUE TITLE
Does salsa delete the data of orphaned structs?

### ISSUE THREAD

**Issue by ScrambledEggsOnToast** [salsa-rs/salsa#438]
Apologies if this is an inappropriate place to ask this question. I'm hoping to use salsa (2022) in a project, and trying to understand whether I will need to do any garbage collection. I'm having a hard time figuring this out from the code/documentation. 

If the database contains data for structs which are no longer used, does that data get automatically deleted?

For context, I have a tree data structure, some of whose nodes are salsa inputs. The shape of the tree is managed by some other functions which occasionally drop some of its branches. Tracked functions are then used to construct a different tree data structure, some of whose nodes are tracked structs --- but these are guaranteed to not depend on the dropped inputs.

**Comment by XFFXFF** [comment]
> If the database contains data for structs which are no longer used, does that data get automatically deleted?

@ScrambledEggsOnToast   Yes, you can assume that Salsa only stores the computation results that are valid for the current [inputs](https://salsa-rs.netlify.app/overview.html#inputs), and all the stale computation results are automatically deleted. 

When you read a field of a tracked struct that is stale, it will panic. You can take a look at the following code snippet as an illustration:
```rust
let input = MyInput::new(&db, 22);
let tracked = tracked_fn(&db, input);

// Modify the input
input.set_field(&mut db).to(24);

// Panic when reading fields of staled tracked structs
tracked.field(&db);
```
This [test](https://github.com/salsa-rs/salsa/blob/master/salsa-2022-tests/tests/panic-when-reading-fields-of-tracked-structs-from-older-revisions.rs) is relevant to this situation

**Comment by ScrambledEggsOnToast** [comment]
@XFFXFF Thanks very much for your reply. The behavior you describe makes sense to me, but one thing that's still not clear to me is whether Salsa keeps around the data of dropped tracked inputs.

Basically:
```rust
{
    let input = MyInput::new(&mut db, 22);
}
// is the data still in the database now?
``` 
This is important to me because I have a tree data structure, with inputs on some nodes, and the shape of the tree can change.

**Comment by XFFXFF** [comment]
> ```rust
> {
>     let input = MyInput::new(&mut db, 22);
> }
> // is the data still in the database now?
> ```
@ScrambledEggsOnToast  Yes, the input data in Salsa is kept around as long as the associated database remains alive. In other words, the data's lifetime is tied to the lifetime of the database. (In fact, the `input` variable in your example is [just an integer](https://salsa-rs.netlify.app/overview.html#salsa-structs-are-just-integers) value that represents a unique identifier for the associated input data)

Here is an example code that demonstrates this behavior:
```rust
    let db = Database::default();

    {
        let _ = MyInput::new(&db, 22);
    }
    let id = salsa::Id::from(0u32);
    let input = MyInput::from_id(id);
    assert_eq!(input.field(&db), 22);
```
Note that using `salsa::Id` directly is not recommended, https://github.com/salsa-rs/salsa/blob/a327acc126d5d760aea83a266b2c5d8c58736003/components/salsa-2022/src/id.rs#L25-L34

**Comment by ScrambledEggsOnToast** [comment]
Hmm, ok. In some scenarios (like mine) this would seem like a memory leak.

How should I remove the old data from the database, and any dependent structs? I guess I could do something like
```rust
{
    let input = MyInput::new(&db, Some(22));

    let tracked = tracked_fn(&db, input);
    // other stuff...

    // Mark an input for deletion. 
    // This will delete all the data of tracked structs 
    // which depend on the input.
    input.set_field(&mut db).to(None);
}
delete_inputs_with_is_none(&db);
```
But I don't actually know how to go about implementing `delete_inputs_with_is_none`. 

I'd also ideally want to wrap `MyInput` in something with an `impl Drop`, but that thing would need a mutable reference to the database, which seems problematic.

**Comment by tv42** [comment]
It's definitely a memory leak. I would expect if you have rust-analyzer running on a tree where source files are created and removed in a loop, it's memory use would increase without bound.

This is unfortunate, and it seems we are saved only by the fact that this tends to be a rare way to use salsa. Historically, the leak was roughly limited by human typing speed.

**Comment by MichaReiser** [comment]
Adding an API to delete any associated memo (cached result) is fairly straightforward. Removing the data stored on the input itself is more involved, especially if we also want to allow re-using the storage location. 

The first would be nice for r-a and ty, but I don't think this is a huge problem because files aren't deleted that frequently. Reusing input slots feels unimportant to me. At least in ty, the `File` struct stores very little data. All heavy data is stored as a query taking `File` as input

**Comment by jorgsowa** [comment]
I also stumbled across this issue. I'm building a language server, and it would be useful to reconsider this API use case. Not a must, but much more flexibility for a consumer.
